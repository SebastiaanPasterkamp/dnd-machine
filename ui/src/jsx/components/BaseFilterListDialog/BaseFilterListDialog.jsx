import React from 'react';
import PropTypes from 'prop-types';
import {
    filter,
    map,
} from 'lodash/fp';

import ModalDialog from '../ModalDialog.jsx';
import utils from '../../utils.jsx';

import './sass/_base-filter-list-dialog.scss';


const BaseFilterListDialog = function({
    items,
    rowComponent: RowComponent,
    label,
    className,
    onCancel,
    onDone,
    onFilter,
    children,
    ...props
}) {
    const style = utils.makeStyle({}, [
        'base-filter-list-dialog',
        className,
    ]);

    const filtered = onFilter
        ? filter(onFilter)(items)
        : items;

    return (
        <ModalDialog
            label={label}
            className={style}
            onCancel={onCancel}
            onDone={onDone}
            subheading={children}
        >
            {map(item => (
                <RowComponent
                    key={item.id}
                    item={item}
                    {...props}
                />
            ))(filtered)}
        </ModalDialog>
    );
};

BaseFilterListDialog.propTypes = {
    items: PropTypes.objectOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
        })
    ).isRequired,
    label: PropTypes.string.isRequired,
    rowComponent: PropTypes.any.isRequired,
    className: PropTypes.string,
    onCancel: PropTypes.func,
    onDone: PropTypes.func,
    onFilter: PropTypes.func,
};

BaseFilterListDialog.defaultProps = {
    className: null,
    onDone: null,
    onFilter: null,
};

export default BaseFilterListDialog;
