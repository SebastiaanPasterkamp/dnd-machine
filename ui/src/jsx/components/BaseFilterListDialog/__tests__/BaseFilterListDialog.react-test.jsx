import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MockRouter from 'react-mock-router';

import BaseFilterListDialog from '../BaseFilterListDialog';

const rowComponent = ({item: {value, id}, ...props}) => (
    <div {...props}>{value}</div>
);

describe('BaseFilterListDialog', () => {
    const fullProps = {
        items: {
            one: {
                id: 1,
                value: 'one',
            },
            two: {
                id: 2,
                value: 'two',
            },
        },
        rowComponent: rowComponent,
        label: "Base Item Picker",
        className: "class-name",
        onCancel: jest.fn(),
        onDone: jest.fn(),
    };

    describe('rendering', () => {
        it('should work with minimum props', () => {
            const tree = renderer.create(
                <MockRouter>
                    <BaseFilterListDialog
                        items={{}}
                        label={fullProps.label}
                        onCancel={fullProps.onCancel}
                        rowComponent={fullProps.rowComponent}
                    />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });

        it('should work with full props', () => {
            const tree = renderer.create(
                <MockRouter>
                    <BaseFilterListDialog
                        {...fullProps}
                        extraProp="passed"
                    />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });

        it('should work with filtered items', () => {
            const tree = renderer.create(
                <MockRouter>
                    <BaseFilterListDialog
                        {...fullProps}
                        onFilter={(item) => item.id !== 1}
                    />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });
    });

    it('should filter items when rendering', () => {
        const onFilter = jest.fn(() => true);

        const wrapper = mount(
            <MockRouter>
                <BaseFilterListDialog
                    {...fullProps}
                    onFilter={onFilter}
                />
            </MockRouter>
        );

        expect(onFilter.mock.calls).toMatchSnapshot();
        expect(onFilter)
            .toBeCalledWith(fullProps.items.one);
        expect(onFilter)
            .toBeCalledWith(fullProps.items.two);
    });

    describe('action buttons', () => {
        const onCancel = jest.fn();
        const onDone = jest.fn();

        const wrapper = mount(
            <MockRouter>
                <BaseFilterListDialog
                    {...fullProps}
                    onCancel={onCancel}
                    onDone={onDone}
                />
            </MockRouter>
        );

        it('should bind onCancel properly', () => {
            expect(onCancel).not.toBeCalled();
            wrapper
                .find('.nice-btn.fa-cross')
                .simulate('click');
            expect(onCancel).toBeCalled();
        });

        it('should bind onDone properly', () => {
            expect(onDone).not.toBeCalled();
            wrapper
                .find('.nice-btn.fa-check')
                .simulate('click');
            expect(onDone).toBeCalled();
        });
    });
});