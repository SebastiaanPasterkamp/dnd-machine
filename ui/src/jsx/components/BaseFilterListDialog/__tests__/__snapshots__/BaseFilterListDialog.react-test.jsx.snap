// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseFilterListDialog rendering should work with filtered items 1`] = `
<div
  className="nice-modal-overlay"
  onClick={[Function]}
>
  <div
    className="modal-dialog nice-modal viewport-center accent tall wide"
    onClick={[Function]}
  >
    <div
      className="nice-modal-content"
    >
      <div
        className="nice-modal-header"
      >
        <a
          className="nice-modal-close"
          onClick={[Function]}
        >
          <i
            className="icon fa-times"
          />
        </a>
        <h4>
          Base Item Picker
        </h4>
      </div>
      <div
        className="nice-modal-body overflow-y-auto overflow-x-hidden"
      >
        <div>
          two
        </div>
      </div>
      <div
        className="nice-modal-footer"
      >
        <div
          className="base-link-group nice-btn-group pull-right"
        >
          <a
            className="cursor-pointer nice-btn icon fa-cross"
            onClick={[Function]}
          >
            Cancel
          </a>
          <a
            className="cursor-pointer nice-btn icon fa-check primary"
            onClick={[Function]}
          >
            Done
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BaseFilterListDialog rendering should work with full props 1`] = `
<div
  className="nice-modal-overlay"
  onClick={[Function]}
>
  <div
    className="modal-dialog nice-modal viewport-center accent tall wide"
    onClick={[Function]}
  >
    <div
      className="nice-modal-content"
    >
      <div
        className="nice-modal-header"
      >
        <a
          className="nice-modal-close"
          onClick={[Function]}
        >
          <i
            className="icon fa-times"
          />
        </a>
        <h4>
          Base Item Picker
        </h4>
      </div>
      <div
        className="nice-modal-body overflow-y-auto overflow-x-hidden"
      >
        <div
          extraProp="passed"
        >
          one
        </div>
        <div
          extraProp="passed"
        >
          two
        </div>
      </div>
      <div
        className="nice-modal-footer"
      >
        <div
          className="base-link-group nice-btn-group pull-right"
        >
          <a
            className="cursor-pointer nice-btn icon fa-cross"
            onClick={[Function]}
          >
            Cancel
          </a>
          <a
            className="cursor-pointer nice-btn icon fa-check primary"
            onClick={[Function]}
          >
            Done
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BaseFilterListDialog rendering should work with minimum props 1`] = `
<div
  className="nice-modal-overlay"
  onClick={[Function]}
>
  <div
    className="modal-dialog nice-modal viewport-center accent tall wide"
    onClick={[Function]}
  >
    <div
      className="nice-modal-content"
    >
      <div
        className="nice-modal-header"
      >
        <a
          className="nice-modal-close"
          onClick={[Function]}
        >
          <i
            className="icon fa-times"
          />
        </a>
        <h4>
          Base Item Picker
        </h4>
      </div>
      <div
        className="nice-modal-body overflow-y-auto overflow-x-hidden"
      />
      <div
        className="nice-modal-footer"
      >
        <div
          className="base-link-group nice-btn-group pull-right"
        >
          <a
            className="cursor-pointer nice-btn icon fa-cross"
            onClick={[Function]}
          >
            Cancel
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BaseFilterListDialog should filter items when rendering 1`] = `
Array [
  Array [
    Object {
      "id": 1,
      "value": "one",
    },
  ],
  Array [
    Object {
      "id": 2,
      "value": "two",
    },
  ],
]
`;
